# RNA-seq analysis with DESeq2 + log-transformed normalized expression values
library(DESeq2)
library(ggplot2)
library(edgeR)
library(factoextra)
library(gplots)
library(RColorBrewer)
countData <- read.csv("01_featurescount/counts.csv", header = TRUE, sep = "\t")
countData <- subset(countData, select=-c(Length))
colnames(countData) <- gsub("Aligned.sortedByCoord.out.bam","",colnames(countData))
metadata <- read.csv("sample_info.csv", header = TRUE, sep = ",")
dds <- DESeqDataSetFromMatrix(countData = countData, colData = metadata, design = ~Type, tidy = TRUE)      
dds <- DESeq(dds)
dds <- estimateSizeFactors(dds)
dds <- estimateDispersions(dds)
dds <- nbinomWaldTest(dds)
res <- results(dds)
dir.create("01_featurescount/02_deseq")
write.table(res, file = "01_featurescount/02_deseq/DEseq_result.csv", sep = "\t", quote = F, col.names = NA)                                     
normCounts = counts(dds,normalized=TRUE)
write.table(normCounts, file = "01_featurescount/02_deseq/normalized_counts.txt", sep = "\t", quote = F, col.names = NA)                       
lognormcounts <- log2(normCounts +1)                                                                           
select = order(rowMeans(lognormcounts), decreasing = TRUE)[1:1000]
hc <- lognormcounts[select, ]
pca <- t(hc)
pca <- prcomp(pca)
group <- as.factor(metadata$Type)
pdf("01_featurescount/02_deseq/RPlots.pdf")
fviz_pca_ind(pca, col.ind = group, palette = c("#00AFBB", "#FC4E07"), addEllipses = TRUE, legend.title = "Group", repel = TRUE)   
res <- results(dds)                                                                                                               
res <- res[order(res$padj), ]                                                                                                    
resdata <- merge(as.data.frame(res), as.data.frame(lognormcounts), by= "row.names", sort=FALSE)                                   
write.csv(resdata, file="01_featurescount/02_deseq/mergedata.csv", quote = FALSE, row.names = F)
select_genes <- rownames(subset(res, padj <0.05))
colnames(lognormcounts) <- c(rep("Bone", 3), rep("Met", 3))
heatmap.2(as.matrix(lognormcounts)[select_genes,], col = colorRampPalette(c("red", "green")), scale = "row", Rowv = TRUE, Colv = TRUE, labRow = "", cexCol = 1.4, trace = "none", dendrogram = "both", density.info = "histogram", margins = c(9,5))
volcanoplot <- function (res, lfcthresh=2, sigthresh=0.05, main="Volcano Plot", legendpos="bottomright", labelsig=TRUE, textcx=1, ...) 
{with(res, plot(log2FoldChange, -log10(pvalue), pch=20, main=main, xlim=c(-3,3)))
with(subset(res, padj<sigthresh), points(log2FoldChange, -log10(pvalue), pch=20, col="blue"))
with(subset(res, abs(log2FoldChange)>lfcthresh), points(log2FoldChange, -log10(pvalue), pch=20, col="orange"))
legend(x= "topright", bg = "gray", legend=c(paste("FDR<",sigthresh,sep=""), paste("|LogFC|>",lfcthresh,sep="")), pch=20, col=c("blue","orange","green"))}
volcanoplot(resdata, lfcthresh=2, sigthresh=0.05, textcx=.8, xlim=c(-3,3))
